src/server.c

```diff
@@ -1685,9 +1685,9 @@ int clientsCronResizeQueryBuffer(client *c) {
     time_t idletime = server.unixtime - c->lastinteraction;
 
     /* There are two conditions to resize the query buffer:
-     * 1) Query buffer is > BIG_ARG and too big for latest peak.
-     * 2) Query buffer is > BIG_ARG and client is idle. */
-    if (querybuf_size > PROTO_MBULK_BIG_ARG &&
+     * 1) Query buffer is > PROTO_RESIZE_THRESHOLD and too big for latest peak.
+     * 2) Query buffer is > PROTO_RESIZE_THRESHOLD and client is idle. */
+    if (querybuf_size > PROTO_RESIZE_THRESHOLD &&
```

Comment:
sorry for dragging this further. i feel a bit uncomfortable.
before #7875, this code thought a newly born client's query buffer is small (32k), and didn't consider releasing it.
#7875 didn't actually increase memory consumption (used_memory or RSS, since we use SDS_NOINIT), but it caused this code to release the buffer quickly.

in this PR, we wanna introduce a change that will retain bigger allocation and avoid releasing them.
i.e. in redis 6.0 we should have released an allocation as soon if it's bigger than 32k, and now we'll avoid releasing it unless it's bigger than 64k.
maybe instead, we need to change the code that allocates the query buffer to avoid the greediness of sdsMakeRoomFor?

@yoav-steinberg @sundb WDYT?

Reply:
This should be the most appropriate, I wanted to do this at first, but sds cannot do this directly.
It needs to be something like the following, but I always find it a bit odd
```
if (sdsalloc(c->querybuf) == 0) {
    c->querybuf = sdsnewlen(c->querybuf, PROTO_IOBUF_LEN);
    sdssetlen(c->querybuf, 0);
} else {
    c->querybuf = sdsMakeRoomFor(c->querybuf, readlen);
}
```

Reply:
Take back the above comment.
I can't think of a way to allocate a fixed 32k size sds.

Reply:
i attempted to add a non-greedy version of sdsMakeRoomFor in the past, it didn't eventually make it in since i found another solution to the problem i was dealing with. but maybe it's time to do that now.
i.e. to be used for the initial size, not for the later growing.

Reply:
Curious to know how you achieved this.

Reply:
maybe i'm missing something...

IIRC i renamed sdsMakeRoomFor to _sdsMakeRoomFor and made it take an `exact` argument.
then i wrapped it with two functions sdsMakeRoomFor that has the original signature, and sdsMakeRoomForExact (which has the same interface, but passes 1 to the `exact` argument.
then the _sdsMakeRoomFor function avoid adding the extra 1mb or *2 when the exact flag is on.

Reply:
i wouldn't get the exact amount i asked for since the internal fragmentation is added (up to 30%), but at least i turn off the 200% extra that the greediness mechanism adds..

Reply:
But there is a problem, like ```sdsMakeRoomFor(s, 32*1024, 1)```, although I want to allocate 32k, but because the hdr of sds is 1 byte more, so the final memory will be 40k, want to allocate 32k need  ```sdsMakeRoomFor(s, 32*1024 - 1, 1)```.

Reply:
You mean ignore the extra memory?
```
usable = usable-hdrlen-1;
if (usable > add_len)
    usable = add_len;
```

Reply:
no, i just mean the greediness of:
```c
    if (newlen < SDS_MAX_PREALLOC)
        newlen *= 2;
    else
        newlen += SDS_MAX_PREALLOC;
```

note that we actually really intended to allocate 16k (`PROTO_IOBUF_LEN`), but because of the greediness we attempt to allocate 32k, which then grows to 40k due to sds header and internal frag.
instead, if we'll really ask to allocate 16k, it'll grow to some 20k with sds header and internal frag.
this will let us bring back PROTO_RESIZE_THRESHOLD to 32k as it was in the past.

Reply:
ohh, this is indeed a good idea, querybuf will be larger than 32k in the second expansion, let me make the change.

Reply:
https://github.com/redis/redis/blob/65851e4562b2ec3da86d0bdd9e5bd5306d28595a/src/networking.c#L1920
I did not modify this expansion, I think since it has been expanded to PROTO_MBULK_BIG_ARG(32k), then let it be bigger to avoid realloc next time.

Reply:
i think i spoke about that in this [comment](https://github.com/redis/redis/pull/9003#discussion_r646103828).
when we detect BIG_ARG, we're gonna copy (re-purpose) that sds to the argv array, and create a new query buffer.
we have no use for the extra space in there. i think that line should be modified to the non-greedy.
am i missing anything?